### ISSUE TITLE
Support other DBMSs (Refactor Proposal)

### ISSUE THREAD

**Issue by gacarrillor** [opengisch/projectgenerator#86]
To support DBMSs other than PostgreSQL/PostGIS, I'd like to refactor Project Generator a bit. Requirements from our side are GeoPackage support and easy addition of other DBMSs like SQL Server.

I would like to know what do you think about handling database connections in a generic way. That is, 

 + Create a `Connection` class with methods like `initialize()` and either `executeSQL( 'query_name', dictParams )` or `particular_sql_query_method( dictParams )` to be overwritten by subclasses.
 + Each new DBMS would subclass the `Connection` class and implement its own way of initializing connections and executing SQL queries. The output of such queries should be compliant with what Project Generator expects (e.g., a list of dictionaries, where each dictionary would correspond to a whole row).
 + Create a `ConnectionManager` that holds the current DBMS so that Project Generator can be aware of this at any time and show proper GUI controls for DB connections as well as use the proper ili2db subproject under the hood.
 + `ConnectionManager` should also be capable of generating connection parameters that ili2db understands.
 + Project Generator would use `ConnectionManager` anytime it needs to access the database, and such class will pass queries to `Connection` subclasses:

```
    self.db = ConnectionManager()
    self.db.setDBMS( 'geopackage' )
    res = self.db.initialize( connectionParams )
    if not res:
        # Show connection problem messages

    # Otherwise...
    tables = self.db.executeSQLQuery( 'get_tables', params ) 
    # Or 
    tables = self.db.get_tables( params )
```

I've implemented this approach in another project and it turned out to be great for our needs of switching between online (PostgreSQL) and offline (SpatiaLite) modes. 

In this way, we wouldn't need Generator classes per DBMS (lot of bussiness logic there) but a single Generator will make use of `ConnectionManager` to deal with each DBMS data handling and structuring.

I think this will make things easier for adding more DBMSs than asking people to write a Generator for the new DBMS (as mentioned above, SQL Server would be the next one to be supported for us).

What do you think about it?

**Comment by m-kuhn** [comment]
If I understand this correctly (I just had a very quick read), the main piece will then be the `Connection` base class, where every connection needs to implement a set of methods like

```
class PostgresConnection(Connection):
    def get_tables(... some parameters ...):
        return list-of-tables

    def get_relations(... parameters ...):
        return list-of-relations

    ... etc
```

Isn't this very similar to the current [PostgresCreator class](https://github.com/opengisch/projectgenerator/blob/master/projectgenerator/libqgsprojectgen/generator/postgres/postgrescreator.py)?

Basically the same will need to be implemented again for other providers, refactoring out code that's identified to be very common into helper methods in the base (`Creator` or `Connection`) class?

**Comment by gacarrillor** [comment]
Connection subclasses would deal with database queries. A Generator would deal with Project Generator dataobjects. In current master, a Generator involves both things.

**Comment by m-kuhn** [comment]
The main trouble at the moment for me seems to be the bits and pieces concerning ili2db metatables which are currently mixed into the pg code, correct?

It seems to me, this is the main thing that needs to be refactored out of the current PostgresGenerator object and instead rely on some hooks (signals or similar) to patch dataobjects created by the Generator.

I like the idea of having a `GeneratorManager` and have an interface to execute arbitrary SQL withing a `Generator`. But I'm not too sure, we can clearly split *queries* from *dataobject generation*. Given that all the databases have different metadata layouts and their field types range from very simple (e.g. sqlite actually doesn't even use static typing, it's spatialite and I think also gpkg that patches column type hints on top). I think the approach to ask the `Generator` *give me all the information you have about schema X* and let it return dataobjects filled as much as possible is not that bad.

It should just be updated a bit differently to *give me all the information you have about schema X and let this/these modifiers (ili2db, yaml, ...) verify and adjust the information before you return it to me*.

```py
class Generator:
    layer_created = pyqtSignal(Layer)
    field_created = pyqtSignal(Layer, Field)

    def addModifier(modifier):
        modifier.setGenerator(self)
        self.modifiers.append(modifier)

    def execute_query(sql):
        ...

    def layers():
        return list-of-Layers

class Ili2DbModifier(Modifier):
    def on_layer_created(layer):
       self.generator.query('SELECT some ili2db metatables')
       layer.setSomeAdditionalAttributes

postgresGenerator = PostgresGenerator()
postgresGenerator.addModifier(Ili2DbModifier())
postgresGenerator.layers()
```

**Comment by gacarrillor** [comment]
Fixed by PR #112. Every new DBMSs should have its corresponding [DBConnector](https://github.com/opengisch/projectgenerator/blob/master/projectgenerator/libqgsprojectgen/dbconnector/db_connector.py) subclass (see [PGConnector](https://github.com/opengisch/projectgenerator/blob/master/projectgenerator/libqgsprojectgen/dbconnector/pg_connector.py) and [GPKGConnector](https://github.com/opengisch/projectgenerator/blob/master/projectgenerator/libqgsprojectgen/dbconnector/gpkg_connector.py)).
